compileopts/target.go

```diff
@@ -56,100 +57,62 @@ type TargetSpec struct {
-		spec.Emulator = spec2.Emulator
-	}
-	if spec2.FlashCommand != "" {
-		spec.FlashCommand = spec2.FlashCommand
-	}
-	if spec2.GDB != "" {
-		spec.GDB = spec2.GDB
-	}
-	if spec2.PortReset != "" {
-		spec.PortReset = spec2.PortReset
-	}
-	if spec2.FlashMethod != "" {
-		spec.FlashMethod = spec2.FlashMethod
-	}
-	if spec2.FlashVolume != "" {
-		spec.FlashVolume = spec2.FlashVolume
-	}
-	if spec2.FlashFilename != "" {
-		spec.FlashFilename = spec2.FlashFilename
-	}
-	if spec2.UF2FamilyID != "" {
-		spec.UF2FamilyID = spec2.UF2FamilyID
-	}
-	if spec2.BinaryFormat != "" {
-		spec.BinaryFormat = spec2.BinaryFormat
-	}
-	if spec2.OpenOCDInterface != "" {
-		spec.OpenOCDInterface = spec2.OpenOCDInterface
-	}
-	if spec2.OpenOCDTarget != "" {
-		spec.OpenOCDTarget = spec2.OpenOCDTarget
+// overrideProperties overrides all properties that are set in child into itself using reflection.
+func (spec *TargetSpec) overrideProperties(child *TargetSpec) {
+	specType := reflect.TypeOf(spec).Elem()
+	specValue := reflect.ValueOf(spec).Elem()
+	childValue := reflect.ValueOf(child).Elem()
+
+	copyFieldIfNotEmpty := func(dst reflect.Value, src reflect.Value) {
+		if src.Len() > 0 {
+			dst.Set(src)
+		}
 	}
-	if spec2.OpenOCDTransport != "" {
-		spec.OpenOCDTransport = spec2.OpenOCDTransport
+
+	copyUintIfNotZero := func(dst reflect.Value, src reflect.Value) {
+		if src.Uint() != 0 {
+			dst.Set(src)
+		}
 	}
-	if spec2.JLinkDevice != "" {
-		spec.JLinkDevice = spec2.JLinkDevice
+
+	copyPtrIfNotNil := func(dst reflect.Value, src reflect.Value) {
+		if !src.IsNil() {
+			dst.Set(src)
+		}
 	}
-	if spec2.CodeModel != "" {
-		spec.CodeModel = spec2.CodeModel
+
+	appendField := func(dst reflect.Value, src reflect.Value) {
+		dst.Set(reflect.AppendSlice(src, dst))
 	}
 
-	if spec2.RelocationModel != "" {
-		spec.RelocationModel = spec2.RelocationModel
+	for i := 0; i < specType.NumField(); i++ {
+		field := specType.Field(i)
+		switch kind := field.Type.Kind(); kind {
+		case reflect.String: // for strings, just copy the field of child to spec if not empty
+			copyFieldIfNotEmpty(specValue.Field(i), childValue.Field(i))
+		case reflect.Uint, reflect.Uint32, reflect.Uint64:
+			copyUintIfNotZero(specValue.Field(i), childValue.Field(i))
+		case reflect.Ptr:
+			copyPtrIfNotNil(specValue.Field(i), childValue.Field(i))
+		case reflect.Slice: // for slices... check if there is a tag for the field
+			if tag, ok := field.Tag.Lookup("override"); ok {
+				switch tag {
+				case "copy":
+					// copy the field of child to spec if not empty
+					copyFieldIfNotEmpty(specValue.Field(i), childValue.Field(i))
+				case "append":
+					// or append the field of child to spec
+					appendField(specValue.Field(i), childValue.Field(i))
+				default:
+					panic("override mode must be 'copy' or 'append' (default). I don't know how to '" + tag + "'.")
+				}
+			} else {
+				// if no tag, then append the field of child to spec
+				appendField(specValue.Field(i), childValue.Field(i))
+			}
```

Comment:
This can be simplified (but also see the comment above):

```suggestion
			switch tag := field.Tag.Get("override"); tag {
			case "copy":
				// copy the field of child to spec if not empty
				copyFieldIfNotEmpty(specValue.Field(i), childValue.Field(i))
			case "append", "":
				// or append the field of child to spec
				appendField(specValue.Field(i), childValue.Field(i))
			default:
				panic("override mode must be 'copy' or 'append' (default). I don't know how to '" + tag + "'.")
			}
```